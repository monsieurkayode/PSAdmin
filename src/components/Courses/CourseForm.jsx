import React from 'react';
import {
  string, shape, bool, arrayOf, func
} from 'prop-types';
import TextInput from '../common/TextInput/Index';
import SelectInput from '../common/SelectInput/Index';

/**
 * @function CourseForm
 *
 * @param {object} props
 *
 * @returns {JSX}
 */
const CourseForm = ({
  title,
  authorId,
  category,
  length,
  errors,
  saving,
  loading,
  authors,
  onChange,
  onSubmit,
  onFocus
}) => (
  <form onSubmit={onSubmit}>
    <TextInput
      name="title"
      label="Title"
      placeholder="React Fundamentals..."
      onChange={onChange}
      loading={loading}
      onFocus={onFocus}
      value={title}
      error={errors.title}
    />
    <SelectInput
      name="authorId"
      label="Author"
      defaultOption="Select Author"
      onChange={onChange}
      onFocus={onFocus}
      value={authorId}
      loading={loading}
      error={errors.authorId}
      options={authors}
    />
    <TextInput
      name="category"
      label="Category"
      placeholder="UI/UX, HTML, Graphql..."
      onChange={onChange}
      loading={loading}
      onFocus={onFocus}
      value={category}
      error={errors.category}
    />
    <TextInput
      name="length"
      label="Length"
      placeholder="5:12"
      onChange={onChange}
      onFocus={onFocus}
      loading={loading}
      value={length}
      error={errors.length}
    />
    <input
      type="submit"
      value={saving ? 'Saving...' : 'Save'}
      disabled={saving}
      className="btn btn-primary"
    />
  </form>
);

CourseForm.defaultProps = {
  title: '',
  authorId: '',
  category: '',
  length: '',
  errors: {},
  authors: []
};

CourseForm.propTypes = {
  title: string,
  authorId: string,
  category: string,
  length: string,
  errors: shape({}),
  authors: arrayOf(shape({})),
  saving: bool.isRequired,
  loading: bool.isRequired,
  onChange: func.isRequired,
  onSubmit: func.isRequired,
  onFocus: func.isRequired
};

export default CourseForm;
