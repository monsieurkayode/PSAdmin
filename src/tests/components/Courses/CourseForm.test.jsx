import React from 'react';
import { shallow } from 'enzyme';
import CourseForm from '../../../components/Courses/CourseForm';
import mockCourse from '../../__mocks__/mockCourse';
import mockAuthors from '../../__mocks__/mockAuthors';

const setup = (saving) => {
  const props = {
    ...mockCourse,
    authors: mockAuthors,
    loading: false,
    saving,
    onSubmit: jest.fn(),
    onFocus: jest.fn(),
    onChange: jest.fn()
  };

  const wrapper = shallow(<CourseForm {...props} />);

  return {
    props,
    wrapper
  };
};

describe('CourseForm Component', () => {
  const { wrapper, props } = setup(false);

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('renders a form element', () => {
    expect(wrapper.find('form').length).toBe(1);
  });

  it('renders three TextInput and a SelectInput component', () => {
    expect(wrapper.find('TextInput').length).toBe(3);
    expect(wrapper.find('SelectInput').length).toBe(1);
  });

  it('renders a form submit button', () => {
    expect(wrapper.find('input[type="submit"]').length).toEqual(1);
  });

  it('form submit button displays Save text when not saving', () => {
    expect(wrapper.find('input[type="submit"]').props().value).toBe('Save');
  });

  it('form submit button displays Saving... text when saving', () => {
    const newWrapper = setup(true).wrapper;
    const saveButton = newWrapper.find('input[type="submit"]');
    expect(saveButton.props().value).toBe('Saving...');
  });

  it('calls onSumit when form is submitted', () => {
    wrapper.find('form').simulate('submit');
    expect(props.onSubmit).toHaveBeenCalled();
  });
});
