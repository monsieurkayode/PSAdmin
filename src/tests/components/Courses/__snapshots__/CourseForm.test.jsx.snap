// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CourseForm Component renders correctly 1`] = `
<form
  onSubmit={[MockFunction]}
>
  <TextInput
    error=""
    label="Title"
    loading={false}
    name="title"
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder="React Fundamentals..."
    value="Building Applications in React and Flux"
  />
  <SelectInput
    defaultOption="Select Author"
    error=""
    label="Author"
    loading={false}
    name="authorId"
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    options={
      Array [
        Object {
          "firstName": "Cory",
          "id": "cory-house",
          "lastName": "House",
        },
        Object {
          "firstName": "Scott",
          "id": "scott-allen",
          "lastName": "Allen",
        },
      ]
    }
    value="cory-house"
  />
  <TextInput
    error=""
    label="Category"
    loading={false}
    name="category"
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder="UI/UX, HTML, Graphql..."
    value="JavaScript"
  />
  <TextInput
    error=""
    label="Length"
    loading={false}
    name="length"
    onChange={[MockFunction]}
    onFocus={[MockFunction]}
    placeholder="5:12"
    value="5:08"
  />
  <input
    className="btn btn-primary"
    disabled={false}
    type="submit"
    value="Save"
  />
</form>
`;
